File: OOP/python/Abstraction/PayrollSystem.py
from abc import ABC, abstractmethod
class Employee(ABC):
    def __init__(self, name, emp_id):
        self.name = name
        self.id = emp_id

    # Abstract method
    @abstractmethod
    def calculate_salary(self):
        pass

    def get_details(self):
        return f"ID: {self.id}, Name: {self.name}"


class FullTimeEmployee(Employee):
    def __init__(self, name, emp_id, monthly_salary):
        super().__init__(name, emp_id)
        self.monthly_salary = monthly_salary

    def calculate_salary(self):
        return self.monthly_salary


class Contractor(Employee):
    def __init__(self, name, emp_id, hours_worked, hourly_rate):
        super().__init__(name, emp_id)
        self.hours_worked = hours_worked
        self.hourly_rate = hourly_rate

    def calculate_salary(self):
        return self.hours_worked * self.hourly_rate


class PayrollSystem:
    @staticmethod
    def show_payroll(employee: Employee):
        print(f"{employee.get_details()} | Salary: ${employee.calculate_salary()}")


# Main execution
if __name__ == "__main__":
    # Abstract reference type usage
    emp1 = FullTimeEmployee("Alice", 101, 5000)
    emp2 = Contractor("Bob", 102, 40, 50)

    PayrollSystem.show_payroll(emp1)
    PayrollSystem.show_payroll(emp2)

File: OOP/java/Abstraction/PayrollSystem.java
abstract class Employee {
    String name;
    int id;

    // Abstract method: Every employee has a salary, but calculation differs
    abstract double calculateSalary();

    public Employee(String name, int id) {
        this.name = name;
        this.id = id;
    }

    public String getDetails() {
        return "ID: " + id + ", Name: " + name;
    }
}

class FullTimeEmployee extends Employee {
    double monthlySalary;

    public FullTimeEmployee(String name, int id, double monthlySalary) {
        super(name, id);
        this.monthlySalary = monthlySalary;
    }

    @Override
    double calculateSalary() {
        return monthlySalary;
    }
}

class Contractor extends Employee {
    int hoursWorked;
    double hourlyRate;

    public Contractor(String name, int id, int hoursWorked, double hourlyRate) {
        super(name, id);
        this.hoursWorked = hoursWorked;
        this.hourlyRate = hourlyRate;
    }

    @Override
    double calculateSalary() {
        return hoursWorked * hourlyRate;
    }
}
public class PayrollSystem {
    public static void main(String[] args) {
        // We use the abstract type 'Employee' for the reference
        Employee emp1 = new FullTimeEmployee("Alice", 101, 5000);
        Employee emp2 = new Contractor("Bob", 102, 40, 50);

        
        showPayroll(emp1);
        showPayroll(emp2);
    }

    public static void showPayroll(Employee e) {
        System.out.println(e.getDetails() + " | Salary: $" + e.calculateSalary());
    }
}